feat(simplify): Provide context option to control simplifications allowed

gwhitney opened the pull request:
Overall, this PR implements a new option to simplify, `context`, which lists properties that various operators should be assumed to have or not have during the simplification process.  The default context allows all previously existing simplifications to occur. Two alternate contexts are provided, `simplify.realContext` and `simplify.positiveContext`, which are guaranteed to preserve the values of expressions on all reals and all positive reals, respectively. Or a custom context with specific operator properties can be passed in, and the simplify rules are tagged with the properties they depend on so they will only be applied if appropriate. Multiple tests for the new option are supplied. (However, even with these tests and the best effort to be complete, it is likely that some rules could still use further annotation, which should hopefully be easy to add if/when simplification anomalies come up.)

Resolves #2391.

This is clearly a larger-than-usual PR. It is the smallest PR that I could come up with that addressed all of the points in the issue it attempts to resolve. However, given that it is large, I am perfectly willing to break it up and submit as a sequence of two or more PRs. If that is preferred, please just let me know what pieces you recommend I break it into. Thanks.

gwhitney commented:
I see the Node 12 and browser test failures. They are both occurring because in my new tests that values are preserved when simplifying, there are some expressions in which both the original and simplified forms produce NaN (those are critical tests; it's exactly simplifications that make NaN become a regular number that realContext is trying to avoid). But in Node 12 and browsers, NaN is not deepEqual to NaN, whereas in Node 14 and Node 16 it seems that it is.

Can you let me know what assertion I can use that will be OK in all settings so that as long as both expressions produce NaN, or both produce the same regular number, or both produce equivalent arrays or objects (which will not be ===), then the assertion will pass? Thanks.

josdejong commented:
Thanks Glen!

I haven't yet looked into the PR itself, but about the NaN issues: so far we just use `isNaN(...)` to test whether something is NaN. We recently had similar issues in #2368 😄 .

gwhitney commented:
Thanks, checking explicitly with isNaN() has cleared up the tests on all platforms.

josdejong commented on src/function/algebra/rationalize.js:
```diff
@@ -177,6 +178,8 @@ export const createRationalize = /* #__PURE__ */ factory(name, dependencies, ({
       const setRules = rulesRationalize() // Rules for change polynomial in near canonical form
       const polyRet = polynomial(expr, scope, true, setRules.firstRules) // Check if expression is a rationalizable polynomial
       const nVars = polyRet.variables.length
+      const optsNoEF = { exactFractions: false }
```
Small thing: maybe rename this to `optsNoExactFractions`? Too much abbreviations make it harder to understand the code 😅.

gwhitney replied:
Check.

josdejong commented on src/function/algebra/simplify.js:
```diff
@@ -403,36 +515,46 @@ export const createSimplify = /* #__PURE__ */ factory(name, dependencies, (
-          newRule = {
-            l: removeParens(parse(rule.l)),
-            r: removeParens(parse(rule.r))
+          newRule = {}
```
Can you extract a little helper function to built up `newRule` here? Like `newRule = createNewRule(...)`? To not distract from the "main" logic here.

gwhitney replied:
Done. I called it "_canonicalizeRule" because it takes a rule object which might take various forms and produces a standard rule object that applyRule can deal with.

josdejong commented on src/function/algebra/simplify.js:
```diff
@@ -454,91 +576,144 @@ export const createSimplify = /* #__PURE__ */ factory(name, dependencies, (
+          if (resNodes === nodes) {
+            resNodes = nodes.slice()
+          }
+          resNodes[i] = newNode
```
I see `resNodes[i] = newNode` here, and earlier `let resNodes = nodes`. So I suppose it is possible that this line will alter the original `nodes` argument. I think it would be good to do a shallow copy from `nodes` into `resNodes` to prevent altering the input arguments.

gwhitney replied:
It is important that `mapRule` not modify `nodes` (i.e., that it returns the === list to nodes) if the rule doesn't alter any of the individual nodes in `nodes`. That's why the code checks each element in turn and only does the shallow copy (via `.slice()` when and if the rule changes one of them. But the shallow copy is already guaranteed to happen if necessary.

josdejong commented on src/function/algebra/simplify/simplifyCore.js:
```diff
@@ -70,14 +67,31 @@ export const createSimplifyCore = /* #__PURE__ */ factory(name, dependencies, ({
+      let simpChild = false
+      let childCount = 0
+      node.forEach(c => {
+        if (++childCount === 1) {
```
I'm personally not a fan of using `++` and `===` in a one liner, I always have to think twice to understand what is happening (and it is easy to introduce an off-by-one bug). Not a big deal, but if ok with you I would prefer splitting it in two lines: first do the increment, next do the equality check.

gwhitney replied:
Split

josdejong commented on src/function/algebra/simplify/util.js:
```diff
@@ -14,56 +14,109 @@ export const createUtil = /* #__PURE__ */ factory(name, dependencies, ({ Functio
+
+  const defaultName = 'defaultF'
+  const defaultContext = {
+    /*      */ add: { trivial: T, total: T, commutative: T, associative: T },
```
nice trick to make these lines better readable 😄

gwhitney replied:
I like to line things up. ;-)

josdejong commented on src/function/algebra/simplify/util.js:
```diff
@@ -14,56 +14,109 @@ export const createUtil = /* #__PURE__ */ factory(name, dependencies, ({ Functio
+      }
     }
-    return commutative[name] || false
+    return defaultContext[defaultName][property]
```
Can it happen that we would return `undefined` here? Is there need for a fallback `... || false`?

gwhitney replied:
Since the default context has a defaultF property which in turn has all of trivial, total, commutative, and associative properties, `hasProperty` can never return `undefined` on any of these properties. But if you call `hasProperty(name, 'transcendental', defaultContext)` you will get `undefined`. But this is arguably a feature, especially since undefined is false-y: when you ask for a property, you can get true or false on known properties, and if you just check the truth value all unknown properties will be false, but if you want to see if it's a property that mathjs knows nothing about, you can specifically check for undefined. This could also potentially help a little with misspelled properties, like 'assocative'.

On the other hand, if you simply want all unknown properties to come back as the exact value `false`, I am happy to add two `|| false` guards in the spots where undefined might come back.

gwhitney replied:
So I have left this as is for now; just let me know if you want the `|| false` put in and I will do it, or feel free to do it yourself if you like.

josdejong replied:
Good point. I like the current approach of returning `undefined` in case of an unknown function, let's go with that 👍

josdejong reviewed (changes requested):
Wow, really impressive piece of work Glen 😎. How you pass the new `context` everywhere is quite straightforward. And I really love how readable and understandable the new `assuming` API reads, like: 
```js
{
  s: '-(c*v) -> v * (-c)', // make non-constant terms positive
  assuming: { multiply: { commutative: true }, subtract: { total: true } }
}
```

I made a couple of minor inline comments, but overall it's a solid piece of work and looks well tested💪 . 

As a last thing: do we need to write a page/section in the documentation explaining how the `context` works on a high level? Maybe on [the Algebra page](https://github.com/josdejong/mathjs/blob/develop/docs/expressions/algebra.md)? Or is the documentation on the `simplify` function itself enough?

gwhitney commented:
Sorry for all of the pushes, I didn't know of any other way to preview changes to "algebra.md" besides pushing it and looking at it in my fork on GitHub. In any case, I think I have addressed all of your comments. Thanks for your encouragement in this (large-ish) change! Hopefully this is ready to merge or nearly so.

EzikeChris commented:
Great Work

josdejong commented:
Thanks for all your updates, this is a solid foundation on which we can expand further 💪 

No problem at all about all the commits, I'll squash it all in one anyway when merging 😁

josdejong commented:
Published in `v10.2.0`. Thanks again.
